pkg/kubelet/dockershim/docker_container_cleanup_manager.go

```diff
@@ -0,0 +1,178 @@
+// giving the IDs of containers that got cleaned up, if any - regardless of whether the cleanup was
+// successful.
+// These channels are mainly intended for tests, and both can be left nil.
+func (cm *containerCleanupManager) start(stopChannel <-chan struct{}, tickChannel chan []string) {
+	if stopChannel == nil {
+		stopChannel = wait.NeverStop
+	}
+
+	wait.Until(func() {
+		cm.mutex.Lock()
+		defer cm.mutex.Unlock()
+
+		containerIDsToCleanup := cm.unsafeContainerIDsToCleanupOnTick()
+
+		for _, containerID := range containerIDsToCleanup {
+			cm.unsafePerformCleanup(containerID)
+		}
+
+		if tickChannel != nil {
+			tickChannel <- containerIDsToCleanup
+		}
+	}, staleContainerCleanupInterval, stopChannel)
+}
+
+// Having those as variables makes them easy to mock in unit tests.
+var (
+	currentNanoTimestampFunc = func() int64 {
+		return time.Now().UnixNano()
+	}
+
+	// The period in between 2 calls to cleanupStaleContainers once started.
+	staleContainerCleanupInterval = 5 * time.Minute
+
+	// Cleanup infos older than this much will be considered stale, and cleaned up.
+	staleContainerCleanupInfoAge = 1 * time.Hour
+)
+
+// If failing to clean up a container that many times, we'll just log an error and forget about this container.
+const maxFailures = 5
+
+// insert allows keeping track of a new container's cleanup info.
+func (cm *containerCleanupManager) insert(containerID string, cleanupInfo *containerCleanupInfo) {
+	if cleanupInfo == nil {
+		return
+	}
+
+	cm.mutex.Lock()
+	defer cm.mutex.Unlock()
+
+	if withMetadata, present := cm.cleanupInfos[containerID]; present {
+		// shouldn't happen, as we only insert at container creation time
+		// but still, let's do the right thing if it does happen, and reset our state
+		klog.Errorf("duplicate cleanup info for container ID %q", containerID)
+
+		withMetadata.cleanupInfo = cleanupInfo
+		withMetadata.timestamp = currentNanoTimestampFunc()
+		withMetadata.failureCount = 0
+	} else {
+		cm.cleanupInfos[containerID] = &containerCleanupInfoWithMetadata{
+			cleanupInfo: cleanupInfo,
+			timestamp:   currentNanoTimestampFunc(),
+		}
+	}
+}
+
+// containerIDsToCleanupOnTick returns the IDs of the container IDs to cleanup at every tick,
+// i.e. those that already have failed at least once, as well as those that have become stale.
+// It assumes that the manager's lock has already been acquired.
+func (cm *containerCleanupManager) unsafeContainerIDsToCleanupOnTick() []string {
+	timestampCutoff := currentNanoTimestampFunc() - int64(staleContainerCleanupInfoAge)
+	containerIDsToCleanup := make([]string, 0, len(cm.cleanupInfos))
+
+	for containerID, withMetadata := range cm.cleanupInfos {
+		if withMetadata.failureCount > 0 || withMetadata.timestamp <= timestampCutoff {
+			containerIDsToCleanup = append(containerIDsToCleanup, containerID)
+
+			if withMetadata.timestamp <= timestampCutoff {
+				klog.Warningf("performing stale clean up for container %q", containerID)
```

Comment:
I think this doesn't have to be warning - since we are cleaning up stale container

Reply:
It shouldn't happen. Normally clean ups should be triggered after the container is successfully started, or removed. Stale clean ups are a sign that something is wrong.